Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,505 @@
+                        message: "Syntax Error: \(digit) cannot be represented in a UInt8"
+                    )
+                }
+                digits.append(digitValue)
+            }
+        } else if decimalString.utf8.count == 1 {
+            guard let digitValue = UInt8(String(decimalString[decimalString.startIndex])) else {
+                throw BSONError.InvalidArgumentError(
+                    message: "Syntax Error: \(decimalString[decimalString.startIndex]) cannot be represented in a UInt8"
+                )
+            }
+            digits.append(digitValue)
+        }
+        return digits
+    }
+
+    private func toString() -> String {
+        // swiftlint:disable:previous cyclomatic_complexity
+        var exponent: Int
+        var significandPrefix: Int
+
+        // If the combination field starts with 0b11 it could be special (NaN/Inf)
+        if self.value.hi.getBits(1...2) == 0b11 {
+            if self.isInfinity {
+                return (self.isNegative ? "-" : "") + "Infinity"
+            }
+            if self.isNaN {
+                return "NaN"
+            }
+            // The number is neither NaN nor Inf
+            // Decimal interchange floating-point formats c,2,ii
+            exponent = Int(self.value.hi.getBits(3...(Self.exponentLength + 2)))
+            significandPrefix = Int(self.value.hi.getBit(20) + 0b1000)
+        } else {
+            // Decimal interchange floating-point formats c,2,i
+            exponent = Int(self.value.hi.getBits(1...Self.exponentLength))
+            significandPrefix = Int(
+                self.value.hi.getBits(
+                    (Self.exponentLength + 1)...(Self.exponentLength + Self.combinationLength)
+                )
+            )
+        }
+
+        exponent -= Self.exponentBias
+
+        var significand128 = UInt128()
+
+        // significand prefix (implied bits) combined with removing the combination and sign fields
+        significand128.hi = UInt64((significandPrefix & 0xF) << 46) | self.value.hi.getBits(18...63)
+        significand128.lo = self.value.lo
+
+        // make a base 10 digits array from significand
+        var significandDigits = [Character]()
+
+        var isZero = false
+
+        if significand128.hi == 0 && significand128.lo == 0 {
+            isZero = true
+        } else if significand128.hi.getBits(0...31) >= 0x20000 {
+            /*
+             * The significand is non-canonical or zero.
+             * In order to preserve compatibility with the densely packed decimal
+             * format, the maximum value for the significand of decimal128 is
+             * 1e34 - 1.  If the value is greater than 1e34 - 1, the IEEE 754
+             * standard dictates that the significand is interpreted as zero.
+             */
+            isZero = true
+        }
+
+        if isZero {
+            significandDigits = ["0"]
+        } else {
+            for _ in 0...3 {
```

Comment:
can we replace the bounds of this for loop with `0...(Self.maxSignificandDigits / 9)`?